### PR TITLE
Problem vs Solution Space Example

### PR DESCRIPTION
**Pull request by NTCoding**
Hello, hello, hello!

On my mission to understand Problem vs Solution space so that I can understand it myself and teach it to others (and understand others when they use it) I triggered some conversations on Twitter today.  Thanks to everyone involved and especially @alexeyzimarev who joined the conversation with a lot of key observations and contributions.

I've now put together another proposal which I believe captures the points we agreed on, and maybe leaves a few open points for discussion.

If there is anything you disagree with or see differently your opinion is always welcome and appreciated. It would be a shame to not finish this discussion now we could be so closed to establishing common ground and a shared understanding.

### CODE REVIEW

**Comment by morcs** [comment]
I really liked another quote from Mathias: "problem space is everything we can't touch, solution space is everything we can make choices in."

**Comment by NTCoding** [comment]
First response on the board. Thanks for the excellent contribution @morcs.

I like the heuristic you quoted and I think it would be great to give examples of what it means so we all have a shared understanding of it.

 Can you apply that heuristic to this PR's example scenario and delineate "everything we can't touch" from "everything we make choices in". Key concepts include:

- unmet user need
- user need solved by an existing solution
- proposed solution
- existing solution
- existing (Sub)domain
- proposed future (Sub)domain (based on proposed solution)
- changes to existing domain
- the teams at Cloud Company and Book Company

If the example is missing concepts or incorrectly defines some concepts, you can call that out too. You're also welcome to provide another concrete scenario if you wish

**Comment by skleanthous** [comment]
While I was not able to participate significantly during the weekend I gave this a lot of thought after our discussion on Friday.

While I already made my views on what is problem vs solution space known ([here](https://twitter.com/SKleanthous/status/1332350897494634496?s=20) and [here in more detail](https://github.com/ddd-crew/strategic-architecture-building-blocks/pull/1#issuecomment-731668568)), I see that some people see some benefit in this (if the tweet and discussion in it is any indication).

Due to this, and because I very much believe we should try to remove confusion and help less experienced people as much as possible, I would like to try and help as much as I can and define these terms in an as clear and understandable way as possible, even if coming with a ubiquitous definition may be impossible. From what I've seen so far, there are a lot of people who have slightly different opinions and definitions of problem vs solutions space. What's worse, some slight differences in definition seem to put different parts of our processes and conceptual pieces into different spaces, causing further confusion (domains vs subdomains for example).

To me, there seem to be one **core** question that no-one seems to even attempt to answer: What is the **value** of clearly defining `problem space` and `solutions space`?

Before attempting to define these terms, imho. we first need to answer the above question. Why? Because:

1. Currently, different people have different definitions
2. The vast majority of the discussion around definitions of the spaces were using phrases like "I see it like this" or "I like this definition". This is all OK; I too have a personal preference :) however...
3. ...if we lack an understanding of the value that we're trying to solve, how are we going to be choosing a definition from the ones we've seen being presented so far?

**Comment by NTCoding** [comment]
Hey Savvas,

Good line of questioning. Here's a brain dump from me. Lot's of caveats about these being raw thoughts and biased by my own personal experiences. I'm happy to hear if others disagree and see things differently (I'm always looking for better models).

I find little value in the problem/solution space model. It doesn't really compile in my head. When you create a solution you create new problems to solve, so your solution space becomes your problem space etc. Then there is the word "problem" which seems specific when we think about opportunities and other reasons to build.

The reason I think it's important to clarify, is because it is a foundational concept in (Strategic DDD) and is required to explain other DDD concepts: Bounded Context is solution space. A Bounded Context should be aligned to a subdomain. A Subdomain is problem space. To make the cliché: "subdomain is problem space, bounded contexts are solution space" useful we need to explain it better or remove it (currently it does more harm than good in IMO).

The landscape metaphor works better for me. We organise our landscape into domains, and all domains can contain problems/opportunities or (possible)solutions.

I don't think I ever use problem space space. I talk about customer segments, value propositions, user needs, jobs to be done, company mission, company goals etc. "What user needs have we identified in the retail domain?".

The problem/solution space is often applied in a very traditional, waterfall sense. We find a problem and build a solution and move onto the next thing. Domain models don't just model existing concepts and processes, most of the domain models I have written have added new concepts to domains that didn't exist before. We are now in a world of continuous delivery, evolutionary design, lean startup, I'm not sure how problem/solution space really fits in anymore.

Hope this is in some way useful at moving the conversation forward.

File: README.md [hunk 1]
```diff
@@ -1,6 +1,37 @@
...
 
-This repository catalogues a language for describing architectural concepts from a Strategic Domain-Driven Design perspective.
+## Problem Space vs Solution Space: Example Scenario
 
-The goal of this repository is to create a shared language that practitioners can apply consistently across projects and domains. Less ambiguity should equally benefit newcomers to DDD and experienced practitioners.
+![Problem_vs_solution_space](problem_vs_solution_space.jpg)
 
+Book Company's mission is to improve every aspect of reading. They have noticed that many readers are frustrated when they forget to take books with them when they leave the house and can't read their books while commuting. 
+
+>Physical Books (subdomain) is part of problem space for Book Company
```
Comment (yellowbrickc) [inline README.md:9]: This is the only phrase, I think is confusing. Strictly speaking, you are right of course but using **everywhere** DDD terms would be (for me at least) counterproductive. I would agree if it would be helpful but what cannot books just be books and nothing bound to DDD. It feels like it would transform a thing easy to understand in something that must be categorized (in a subdomain in this case). And opens the channel for other (not helpful) questions like "are you sure that books are subdomains and not domains?"

**Review by yellowbrickc — COMMENTED** [review]
I think it is cool, but the one comment I made about the books. It is not a blocker of course just an opinion.
 
Regarding the explanations, I would also include a sentence explaining the benefits of these 2 terms. Something like
"Using these terms can help during discussions about (alleged) requirements because it sets the focus on the needs and stops people from jumping to (hastily made) decisions regarding the solution before understanding the real problem"

File: README.md (hunk 1 above)
Comment (NTCoding) [reply README.md:9]: Thanks for the feedback @yellowbrickc.

Can I infer from your comment that you are saying that there are no domains in the problem space? I.e. physical books is our problem space therefore it cannot be a (sub)domain?

On the topic of: 

"And opens the channel for other (not helpful) questions like "are you sure that books are subdomains and not domains?""

This unfortunately is the cost of fuzziness. We like terms like domain and subdomain because they are fuzzy and can be applied in new and novel scenarios. But as a trade-off, it's hard to pin down precisely what they mean.

And just to clarify, with this diagram, I am just trying to visualise how I think each part of the system would be labelled according to DDD terminology. In DDD, the subdomain (physical books) are part of the problem space. I wouldn't use this diagram to teach DDD because I don't think problem/solution space are clear enough to be uuseful.

If you get chance, I would love to see you could create a version of this diagram labelling each part how you interpret DDD. Then we can see where we have common ground and where we interpret parts of DDD differently (open invitation to everyone else as well).

Comment (yellowbrickc) [reply README.md:9]: @NTCoding 

>  there are no domains in the problem space

Just like Mathias says "what's a solution space for one, is the problem space for another" I just don't think that it helps by the clarification. I see it like @skleanthous, I would try to leave out everything creating even more fuzziness. I do not think, that we should see everything in DDD terms. These terms are not originated in DDD, they just used a lot by people in DDD (IMO).

I'm waiting (eagerly) for the end of the year and my holiday (meaning not working) and I hope that I will find the energy to create a fitting diagram but as I said: I agree with the rest, I think it describes well the situations.

**Comment by yellowbrickc** [comment]
I have no idea how I could oversee the comments from @skleanthous, sorry for this. Must be a combination of long working hours and some secret github feature.

I admire the effort you are making Savvas in helping to explain something what you don't even use 👍 

I have the same intention as you: helping to explain what are the benefits of these terms, how and when can they be helpful. In order to achieve this, I propose adding the sentence 
>Using these terms can help during discussions about (alleged) requirements because it sets the focus on the needs and stops people from jumping to (hastily made) decisions regarding the solution before understanding the real problem 

if we all agree.

Who are this "all" and how we solve point 3 from your list of questions is of course still open. Just like in the other PR, I have some problems with long discussions without deciding anything and acting correspondingly. 

The only solution I can imagine in this Open Source repo is to go for the compromise and not for the veto. Additionally, as Nick was the one who created these repos, he can make the last decision. This works well with software because the maintainer knows the intent and has his/her own ideas about the design. I wasn't able to find out how the "decision processes" were planed in these repos though so the only thing I can do is to propose this and hope that we start with a consensus here 🙂 

After a week-long discussion on Twitter, I'm sure that we won't get THE ultimate truth. I think it will be still very helpful for others is we describe the context and the benefits of all these terms.

**Comment by skleanthous** [comment]
Hey @yellowbrickc, I may not use these terms any longer, but I did struggle quite a lot with them since a long time ago. As I mentioned on the other thread, I spent quite a lot of time this year trying to understand these terms well. I did a lot of research (I hope this was evident by my explanation of the mathematical definition) before deciding to stop using them. I also had quite a lot of bad experiences as I described on my comment that I linked to above (for ref: https://github.com/ddd-crew/strategic-architecture-building-blocks/pull/1#issuecomment-731668568). I personally experienced all of 1-3 points on more than one occasion. None the less, I'm fully prepared to adopt them again if the community sees value in them and we can clearly articulate it, and be able to explain it to people so that we can all understand the same thing.

Coming back to the point about problem vs solution space. 

On the comment from @yellowbrickc, I do agree fully that we don't need to exclusively use DDD terms to define our language or to do analysis. First of all, (strategic) DDD is one of many tools we can use, and I don't think it's a good idea to exclusively see things through DDD-tinted glasses. Then, let's not forget that even from DDD, the UL we establish for our BC's is a _subset_ of the language the domain experts use when they communicate. The UL contains the domain-specific terms the domain experts use while still speaking a human language.

From @NTCoding reply: 

> Bounded Context is solution space ... A Subdomain is problem space

Personally, I can't even see how or why using these terms actually **helps** convey this mesage? As I said on the other thread, this just feels like taxonomy to me, rather than providing context to anyone who hears this. The context can be conveyed using terms that are easier to understand, without losing meaning - on the contrary. As I see it, context is in fact much better served by being more specific on the intent, and I think Nick puts it quite nicely later in his reply:

> I talk about customer segments, value propositions, user needs, jobs to be done, company mission, company goals

Regardless, I am open to hear from all the other people who use this term and try to understand why they do so and what is the benefit they gain from it. Most of the twitter responses revolves around asking people to focus on the problem space instead of focusing on solution, which is a very specific benefit imho, with good alternatives. Which brings me to the comment by @yellowbrickc:

> Using these terms can help during discussions about (alleged) requirements because it sets the focus on the needs and stops people from jumping to (hastily made) decisions regarding the solution before understanding the real problem

I don't see how saying `let's focus on the problem space` is more beneficial than explaining `we need to stop jumping to decisions or solutions before understanding the real problem`, as it was so nicely put. The second is easier to understand and conveys intent in a _much_ better way imo.

As a last point, I think having a documented discussion like this, where we can point people to, with an explanation of the values we see in these terms (if any) along with how we ended up defining them is useful. As long as there's something writen down, we can point people to it, and potentially slowly influence the way people speak.

**Comment by alexeyzimarev** [comment]
> Personally, I can't even see how or why saying that actually **helps**?

I think it helps a lot. It helps, first of all, with clarity. Bounded context has nothing to do with business context. It is, by definition, something that already exists in software. It is a part of the implementation. Therefore, it can be wrong, misplaced and misunderstood. Most of the software out there has a single bounded context, implemented as a big ball of mud.

Domains have nothing to do with software, _unless_ the software is the domain. For most domains out there, if you remove all the software (and, therefore, all the bounded contexts), those domains will continue to exist.

A lot of struggle in the industry comes from people in tech not willing or not being able to understand _the domain_ as the **problem space** and create software as the **solution** for non-existing problems or software that doesn't solve the problem. This disconnection is what DDD aims to solve, essentially.

That's why I believe that making such a clear distinction really help.s

**Comment by skleanthous** [comment]
Hello @alexeyzimarev. I was not trying to comment about the meaning of the sentence `Bounded Context is solution space ... A Subdomain is problem space` I was commenting that it's not useful to explain it in that terms vs something else. More specifically, I don't see a benefit saying `BC is in the solution space` compared to your explanation of `Bounded context has nothing to do with business context. It is, by definition, something that already exists in software. It is a part of the implementation` which is very clear, specific and easy to understand.

I'll update my earlier comment to reflect that more clearly.

To be clear, I agree _fully_ when you say:

> A lot of struggle in the industry comes from people in tech not willing or not being able to understand the domain as the problem space and create software as the solution for non-existing problems or software that doesn't solve the problem. This disconnection is what DDD aims to solve, essentially.

I am coming from this exact point of view. All I am trying to do is make sure that we can all communicate better, and people to understand the same thing. And to take this a bit further, a lot of people who are active in this community actually have a pretty close understanding of these terms by the mere act of interacting on a semi-frequent basis. The problem is that new people trying to enter this area are met with a wall of vague, and ambiguous terms. We only stand to benefit as a community by removing that ambiguity and making this field easier for people to enter.

**Comment by skleanthous** [comment]
To be more clear, and for people coming in to this discussion, there is nothing wrong with _using_ the problem and solution space concepts or the spaces themselves (on the contrary there's quite a lot of value to be had there). The problem is also not with the meaning. The issue is with the _language_ we use to refer to them and the terms themselves. It's not clear for everyone what we mean when we use these terms, and people understand slightly different (and sometimes conflicting) things.

**Comment by yellowbrickc** [comment]
@skleanthous 
I've just spoken this loud out
>I don't see how saying let's focus on the problem space is more beneficial than explaining we need to stop jumping to decisions or solutions before understanding the real problem, as it was so nicely put. The second is easier to understand and conveys intent in a much better way imo.

and I realized that you are partially right, normally I use both 🙂 I use something like the second sentence to stop the "slide" and afterwards the terms to label the 2 sides of the boundary.

**Comment by yellowbrickc** [comment]
I love this @alexeyzimarev 
> A lot of struggle in the industry comes from people in tech not willing or not being able to understand the domain as the problem space and create software as the solution for non-existing problems or software that doesn't solve the problem. This disconnection is what DDD aims to solve, essentially.

Can we use this as an intro for the repo @NTCoding ?

**Comment by NTCoding** [comment]
> I love this @alexeyzimarev
> 
> > A lot of struggle in the industry comes from people in tech not willing or not being able to understand the domain as the problem space and create software as the solution for non-existing problems or software that doesn't solve the problem. This disconnection is what DDD aims to solve, essentially.
> 
> Can we use this as an intro for the repo @NTCoding ?

Hey @yellowbrickc,

I think it would be great if we had an example laid out using this definition of problem space, since it is different to the one proposed at the start of this PR.

We would need to make it clear how: domain & subdomain are problem space and a 1:1 between subdomains (problem space) and bounded contexts (solution space) is recommended. If a bounded context is the solution space, how can it align with something in the problem space when the solution is creating things that don't yet exist? I'm not saying it's wrong, I just don't understand it, so I'd be grateful for examples.

You might also like the example from Zhamak Dehghani where she talks about data domains https://martinfowler.com/articles/data-mesh-principles.html. Maybe you could use that example to point out what your definition of problem space and solution space are?

There's a good quote [in this article](https://fs.blog/2015/01/richard-feynman-knowing-something/) from the works of Feynman: “Knowing the name of something doesn’t mean you understand it. We talk in fact-deficient, obfuscating generalities to cover up our lack of understanding.“ I feel that's where we are it with the problem/solution space and it's time to not speak in generalities and use concrete examples to show we understand it.

**Comment by yellowbrickc** [comment]
I will try to do this at the weekend. 

Last words for today: I think the biggest "translation issue" we have here lies in the word _domain_: for me, **domain** in the quote from Alexey means **business** but this could be because I'm using that word a lot in that sense. In my view, the only DDD term used in that intro is the acronym DDD which would be good for an intro, a reason for the whole repo.

Am I wrong @alexeyzimarev?

**Comment by alexeyzimarev** [comment]
That's right, this is what I mean. Looking in the dictionary, the domain is a specific area of activity or knowledge, whilst Eric used a more specific definition in the DDD Reference:

> Domain: a sphere of knowledge, influence, or activity. The subject area to which the user applies a program is the domain of software.

So, we write software **solutions** to solve **problems** of the domain.

**Comment by alexeyzimarev** [comment]
@skleanthous 
> I don't see how saying let's focus on the problem space is more beneficial than explaining we need to stop jumping to decisions or solutions before understanding the real problem

_Let's focus on the problem space_: is very neutral. very short and calls to action (focus).

_We need to stop jumping to decisions or solutions before understanding the real problem_: can be seen as passive-aggressive as saying that presumes that someone here (we) already are jumping to solutions and don't even understand something. It is also calling _not_ to act (stop), without proposing an alternative.

**Comment by skleanthous** [comment]
@alexeyzimarev I copied that example (those were not my words if you read my message and thread carefully), trying to demonstrate that the second was far more clear in its intent. And it is. If the intent is to stop people veering off to trying to solve a problem in the midst of a workshop to get context, why not just say so? Be polite for sure, but be explicit in what you want people to do, or stop doing.

Also, being vague on purpose runs a higher risk of misunderstanding and conflict when we understand different things.

Basically, all I'm trying to say (from what I see something I've not done that well so far) is that if one needs to tell people to focus on the purpose of a workshop, say it, explicitly. Don't say "focus on solution space" and leave people to interpret that from their own understanding.

**Comment by yellowbrickc** [comment]
Hi everyone,

this is how I see and use these 2 terms. (The miro board should be public as I only have a free account)

Problem space - solution space
https://miro.com/app/board/o9J_lcJDXIY=/

Happy to get your feedback 🙂

**Comment by yellowbrickc** [comment]
I must add an urgent disclaimer to this, I have a problem with designing solutions without having the chance to understand and challenge the requirements 😆 
Please get all this as fictional, with a huge "maybe" everywhere in the solution space. For example, I labelled the localizer as a bounded context but now, after less than 5 minutes I would like to change this and label it as a service in the subdomain **Reminders**. How something is organized depends a lot on the size of the IT department, the goal and the strategy of the company, etc

**Comment by NTCoding** [comment]
> Hi everyone,
> 
> this is how I see and use these 2 terms. (The miro board should be public as I only have a free account)
> 
> Problem space - solution space
> https://miro.com/app/board/o9J_lcJDXIY=/
> 
> Happy to get your feedback 🙂

Hi @yellowbrickc,

Thanks for doing this. A very neatly laid out example, and I love how you structured the problem space.

I think your proposal makes sense and is logically sound. The only thing that I would say doesn't align to other definitions of problem space is that the proposed solutions (needing reminders, and having their books on an app) are in the problem space when they are proposed solutions.

Also, this definition contradicts with the DDD notion that subdomains are problem space. For the record I prefer your model because it seems logically coherent (whereas subdomains in problem space I cannot make sense of).

**Comment by NTCoding** [comment]
@yellowbrickc 

"I would like to change this and label it as a service in the subdomain Reminders. How something is organized depends a lot on the size of the IT department, the goal and the strategy of the company, etc"

There's a nice polarity map in here for @Baasie: the polarities of fuzzy and highly-constrained definitions.

**Comment by yellowbrickc** [comment]
I enabled [discussions](https://github.com/ddd-crew/strategic-architecture-building-blocks/discussions/) for this repo, it could be helpful I think 🙂 
Generally, I would be happy if we could come to a common understanding and merge this PR soon (it feels somehow as an open action point)

**Comment by NTCoding** [comment]
@yellowbrickc This PR was mainly just to move the discussion forward. I've closed it now because it looks like the conversation has ended. I don't think we need to merge the changes. I'm going to publish something like this on my personal github which reflects my view/definitions. Maybe in the future when we gain some consensus on problem/solution space we can try again in DDD Crew. Thanks for all the great comments :)

**Comment by carousel** [comment]
Some suggestions for future meetups: 

- importance of culture when modeling 
- can we have ubiquitous language as visual (or other types) not just spoken or verbal?
- expressiveness and limitations of our tools/programming language to solve domain problem
- is software design dead?
- AI and DDD

**Comment by yellowbrickc** [comment]
@carousel 
We could continue the discussion here https://github.com/ddd-crew/strategic-architecture-building-blocks/discussions/5#discussion-2310976
